turnpike/plugins/vpn.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import re
+
+from http import HTTPStatus
+from flask import request
+
+from ..plugin import TurnpikePlugin
+
+
+class VPNPlugin(TurnpikePlugin):
+    vpn_pattern = r"(mtls\.|)private\.(console|cloud)\.(stage\.|dev\.|)redhat\.com"
+    edge_host_header = "x-rh-edge-host"
+    vpn_config_key = "private"
+
+    def __init__(self, app):
+        self.vpn_regex = re.compile(self.vpn_pattern)
+        self.env = app.config.get("WEB_ENV").casefold()
+        self.headers_needed = set(self.edge_host_header)
+
+        super().__init__(app)
+
+    def process(self, context):
+        self.app.logger.info("env: %s", self.env)
+        if self.vpn_config_key not in context.backend or context.backend[self.vpn_config_key] != True:
+            return context
+
+        edge_host = request.headers.get(self.edge_host_header)
+        backend_name = context.backend["name"]
+
+        if not edge_host:
+            # TODO: integrate glitchtip with turnpike and capture this so we get alert if it happens, see https://issues.redhat.com/browse/RHCLOUD-40788
+            return self.forbidden(
+                context,
+                logging.WARNING,
+                "request to backend '%s' denied - missing '%s' header which is required for vpn restricted backend",
+                backend_name,
+                self.edge_host_header,
+            )
+
+        match = self.vpn_regex.fullmatch(edge_host)
+
+        if not match:
+            return self.forbidden(
+                context,
+                logging.DEBUG,
+                "request to backend '%s' denied - '%s':'%s' does not originate from vpn restricted edge host",
+                backend_name,
+                self.edge_host_header,
+                edge_host,
+            )
+
+        match_env = match.groups()[2]
+        if self.is_production() and match_env != "":
+            return self.forbidden(
+                context,
+                logging.INFO,
+                "request to backend '%s' denied - '%s':'%s' is from edge host in wrong env, expected prod host",
+                backend_name,
+                self.edge_host_header,
+                edge_host,
+            )
+        elif not self.is_production() and match_env == "":
+            return self.forbidden(
+                context,
+                logging.INFO,
+                "request to backend '%s' denied - '%s':'%s' is from edge host in wrong env, expected non prod host",
+                backend_name,
+                self.edge_host_header,
+                edge_host,
+            )
```

Comment:
If you accept the regex suggestion, this would change a bit, since we only would have one `Optional[str]` resulting group:

```suggestion
        match_env = match.groups()[0]
        if self.is_production() and match_env:
            return self.forbidden(
                context,
                logging.INFO,
                "request to backend '%s' denied - '%s':'%s' is from edge host in wrong env, expected prod host",
                backend_name,
                self.edge_host_header,
                edge_host,
            )
        elif not self.is_production() and not match_env:
            return self.forbidden(
                context,
                logging.INFO,
                "request to backend '%s' denied - '%s':'%s' is from edge host in wrong env, expected non prod host",
                backend_name,
                self.edge_host_header,
                edge_host,
            )
```

Reply:
updated this since I applied the regex suggestion